invokeai/app/services/image_record_storage.py

```diff
@@ -204,6 +206,13 @@ def _create_tables(self) -> None:
                 """
             )
 
+        if "workflow" not in columns:
+            self._cursor.execute(
+                """--sql
+                ALTER TABLE images ADD COLUMN workflow TEXT;
```

Comment:
I feel like eventually we will need a workflows table that this can reference - any reason not to go ahead and do it that way? It would make the migration path easier

Reply:
Hmm. We will eventually have two kinds of workflows:
1. workflows associated with images (every generated image will have one eventually, and uploaded images may have one)
2. library workflows (created when the user clicks save workflow)

My thought was that these are qualitatively different things. 

#1 is image metadata
#2 is its own entity

If they are on the same table, it means any list queries for library workflows will always, 100% of the time, use a constraint to exclude all the type 1 workflows (associated with images).

Given the handling of these two kinds of workflows is so different, I think it makes sense for type 1 workflows to be stored alongside the image and type 2 to be in a separate table (future pr).

Users would be able to save an image's workflow to their workflow library, copying it from the images.workflow column into the library workflows table.

Reply:
I think that if we were designing workflow library at this time, I would think a `workflows` table that images are associated to makes sense, and maybe a boolean `isSaved` to mark the ones that a user wants to reuse. It just doesn't make a lot of sense to have two different ways to save what is kind of the same thing. I also think it would be nice to reduce the size of these large image DB entries. In this world of two distinct types of workflows, f an image is generated with a saved workflow, how would we know? What if someone wants to filter by workflow used?

Just playing devils advocate, I can definitely see your argument and I know its a PITA to change SQLite schemas after the fact but it is doable if we made a barebones workflow table for now that we built out more as teh library developed

Reply:
I was concerned about needing to create columns for an evolving workflow schema - SQLite json parsing is slow so I though we'd need to split out cols for each workflow attribute. 

But I discovered that SQLite supports indexed expressions.

You can use the SQLite function `json_extract()` to create an index on a json key! Combined with generated columns, SQLite can function very similarly to document database.

The best part is, you can start with just a single json column and incrementally add generated columns with indices as needed. 

I'll implement the workflows table like this.

I also discovered SQLite has a built in very-fast full text search which may prove very useful.